pkg/distributor/ha_tracker.go

```diff
@@ -0,0 +1,312 @@
+	// a prefix creates a ring.PrefixClient, we save the prefix value and set it to
+	// empty string on the config.
+	prefix := cfg.KVStore.Consul.Prefix
+	cfg.KVStore.Consul.Prefix = ""
+
+	switch cfg.KVStore.Store {
+	case "consul":
+		client, err = ring.NewConsulClient(cfg.KVStore.Consul, codec)
+		if err != nil {
+			return nil, err
+		}
+	case "inmemory":
+		// TODO: do we want to use the same init do for sync.Once as we do in the ring?
+		client = ring.NewInMemoryKVClient(codec)
+	default:
+		return nil, fmt.Errorf("invalid KV store type: %s", cfg.KVStore.Store)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t := &tracker{
+		logger:           util.Logger,
+		cfg:              cfg,
+		elected:          map[string]replicaTracker{},
+		writeTimeout:     10 * time.Second.Nanoseconds(),
+		overwriteTimeout: time.Minute.Nanoseconds(),
+		prefix:           prefix,
+		client:           client,
+		cancel:           cancel,
+	}
+	go t.loop(ctx)
+	return t, nil
+}
+
+// Follows pattern used by ring for WatchKey.
+func (c *tracker) loop(ctx context.Context) {
+	defer close(c.done)
+	c.client.WatchPrefix(ctx, c.prefix, func(value interface{}) bool {
+		if value == nil {
+			return true
+		}
+		return true
+	})
+}
+
+// Stop ends calls the trackers cancel function, which will end the loop for WatchPrefix.
+func (c *tracker) Stop() {
+	c.cancel()
+	<-c.done
+}
+
+func (c *tracker) setTimeout(overwrite, write int64) {
+	c.overwriteTimeout = overwrite
+	c.writeTimeout = write
+}
+
+// Wraps calls to client CAS with locking and error checking.
+func (c *tracker) casWrapper(ctx context.Context, cluster, replica string, now int64) error {
+	err := c.client.CAS(ctx, fmt.Sprintf("%s/%s", c.prefix, cluster), func(in interface{}) (out interface{}, retry bool, err error) {
+		desc, ok := in.(*ReplicaDesc)
+		// There was either invalid or no data for the key, so we now accept samples
+		// from this replica.
+		if !ok || desc == nil {
+			desc = &ReplicaDesc{
+				Replica:   replica,
+				Timestamp: now,
+			}
+			return desc, false, nil
+		}
+
+		// As far as the timestamp in Consul is concerned none of the Distributors have
+		// received a sample from the stored replica recently, so we should overwrite the
+		// replica in Consul and accept this sample.
+		if now-desc.Timestamp > c.overwriteTimeout {
+			desc = &ReplicaDesc{
+				Replica:   replica,
+				Timestamp: now,
+			}
+			return desc, false, nil
+		}
+
+		// TODO: resolve this below
+		// I think we want to get rid of this block. The CAS should only write to Consul if necessary, and then return
+		// the replica desc for the replica we want to accept samples from, or nil.
+		// Or should we return a new error type.
+		if desc.Replica != replica {
+			level.Info(util.Logger).Log("msg", "replicas did not match in CAS call")
+			return nil, false, fmt.Errorf("replicas did not match, rejecting sample: %s != %s", replica, desc.Replica)
```

Comment:
I need some time to think about what we should be doing here. I want to make sure we return a value that indicates that we shouldn't actually CAS (returning desc would do that), but we also want to get the value of desc out of this and store it locally. 

Right now I don't know how if we're accepting samples from replica2, but the current distributor doesn't know that yet, would make the earlier call to consul to ask it what replica we should accept from and save that information locally but still reject the sample from replica1. Other than getting `WatchPrefix` actually working.

Reply:
I think returning an error here is fine, but it should be `httpgrp.Errorf(http.StatusOK, message...)`.